GAE/src/org/waterforpeople/mapping/dataexport/RawDataSpreadsheetImporter.java

```diff
@@ -136,17 +126,81 @@ public void runImport(File file, String serverBase, Map<String, String> criteria
         Map<Long, Long> columnIndexToQuestionId = processHeader(sheet);
         Map<Long, QuestionDto> questionIdToQuestionDto = fetchQuestions(serverBase, criteria);
 
+        List<InstanceData> instanceDataList = parseSheet(sheet, questionIdToQuestionDto,
+                columnIndexToQuestionId);
+
+        List<String> importUrls = new ArrayList<>();
+
+        for (InstanceData instanceData : instanceDataList) {
+            String importUrl = buildImportURL(instanceData, criteria.get("surveyId"),
+                    questionIdToQuestionDto);
+            importUrls.add(importUrl);
+        }
+
+        for (String importUrl : importUrls) {
+            invokeUrl(serverBase, importUrl, true,
+                    criteria.get(KEY_PARAM));
+        }
+        Iterator<Row> rowIterator = sheet.rowIterator();
+        int rowCount = 0;
+        while (rowIterator.hasNext()) {
+            rowIterator.next();
+            rowCount++;
+        }
+        // now update the summaries
+        if (rowCount * questionIdToQuestionDto.size() < SIZE_THRESHOLD) {
+            invokeUrl(serverBase, "action=" + RawDataImportRequest.UPDATE_SUMMARIES_ACTION + "&" +
+                    RawDataImportRequest.SURVEY_ID_PARAM + "=" + surveyId, true,
+                    criteria.get(KEY_PARAM));
+        }
+        invokeUrl(serverBase, "action=" + RawDataImportRequest.SAVE_MESSAGE_ACTION + "&" +
+                RawDataImportRequest.SURVEY_ID_PARAM + "=" + surveyId, true,
+                criteria.get(KEY_PARAM));
+
     }
 
     /**
      * Parse a raw data report file into a list of InstanceData
-     *
-     * @param file
+     * 
+     * @param sheet
+     * @param columnIndexToQuestionId
+     * @param questionIdToQuestionDto
      * @return
      */
-    public List<InstanceData> parseSheet(Sheet sheet) throws Exception {
+    public List<InstanceData> parseSheet(Sheet sheet,
+            Map<Long, QuestionDto> questionIdToQuestionDto, Map<Long, Long> columnIndexToQuestionId)
+            throws Exception {
+
+        List<InstanceData> result = new ArrayList<>();
+        int md5Column = Collections.max(columnIndexToQuestionId.keySet()).intValue();
```

Comment:
We should also probably take into account the possibility of the last question being a `GEO` question and therefore the `MD5` column will be later.